[MRG] MAINT Use list and dict comprehension

rth opened the pull request:
Before deprecating Python 2.7 we should probably make sure we are using most of its useful features.

This uses list (and dict) comprehension when possible in the cases where it doesn't hurt readability.

rth committed:
Use list comprehension

rth committed:
Use dict comprehension

rth commented:
It has nothing to do with dropping Python 2 -- my point was that we don't systematically migrate legacy code to use all the features of the latest supported version, here python 2.7 and dict comprehension.

The motivation is that list and dict comprehension provides a more concise way to create lists/dics that also happen to be faster (not that it matter too much here). This doesn't change functionality.

Because contributors often adapt existing code when adding functionality (see e.g. https://github.com/scikit-learn/scikit-learn/pull/12319#discussion_r224973974) , I think it is important not to use anti-patterns (e.g. https://docs.quantifiedcode.com/python-anti-patterns/readability/not_using_a_dict_comprehension.html ) to break this cycle.

rth commented:
Compare,
```py
        self.lambdas_ = []
        for col in X.T:
            with np.errstate(invalid='ignore'):  # hide NaN warnings
                lmbda = optim_function(col)
                self.lambdas_.append(lmbda)
        self.lambdas_ = np.array(self.lambdas_)
```
in `PowerTransformer._fit` to
```py
        with np.errstate(invalid='ignore'):  # hide NaN warnings
            self.lambdas_ = np.array([optim_function(col) for col in X.T])
```
(though granted it's probably one of the most favorable examples).

qinhanmin2014 commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -760,14 +756,10 @@ def test_predefinedsplit_with_kfold_split():
-    for train_ind, test_ind in ps.split():
-        ps_train.append(train_ind)
-        ps_test.append(test_ind)
+    ps_train, ps_test = zip(*ps.split())
```
Maybe keep the original version here?
It's easier to understand and more consistent with the definition of kf_train/kf_test (and I think we'll get two tuples instead two lists here?)

qinhanmin2014 reviewed (approved):
I think it's (slightly) faster and more readable.

eamanu commented:
I agree with @rth and @qinhanmin2014, this changes are faster and the contributors (and new contributors like me) appreciate that because we use existing codes to write new codes,  and this way,  the new codes are more readeable for both contributors and reviewers.

jnothman commented on sklearn/ensemble/forest.py:
```diff
@@ -313,11 +313,9 @@ def fit(self, X, y, sample_weight=None):
-                tree = self._make_estimator(append=False,
-                                            random_state=random_state)
-                trees.append(tree)
+            trees = [
```
We generally prefer visual indentation such as:
```py
            trees = [self._make_estimator(append=False,
                                          random_state=random_state)
                     for i in range(n_more_estimators)]
```
(This may not always be preferable when there is a lot to the left of `[`, but a bare `= [` looks ugly)

rth replied:
>  but a bare = [ looks ugly)

Updated as suggested, though using https://github.com/ambv/black might change your mind about this :)

jnothman commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -1167,14 +1167,12 @@ def fit(self, X, y):
-                alphas.append(_alpha_grid(
-                    X, y, l1_ratio=l1_ratio,
-                    fit_intercept=self.fit_intercept,
+            alphas = [
```
prefer
```py
            alphas = [_alpha_grid(X, y, l1_ratio=l1_ratio,
                                  fit_intercept=self.fit_intercept,
                                  eps=self.eps, n_alphas=self.n_alphas,
                                  normalize=self.normalize, copy_X=self.copy_X)
                      for l1_ratio in l1_ratios]
```

jnothman commented on sklearn/multioutput.py:
```diff
@@ -563,9 +563,9 @@ def fit(self, X, Y):
-        self.classes_ = []
-        for chain_idx, estimator in enumerate(self.estimators_):
-            self.classes_.append(estimator.classes_)
+        self.classes_ = [
```
my preference:
```py
        self.classes_ = [estimator.classes_
                         for chain_idx, estimator
                         in enumerate(self.estimators_)]
```

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2369,10 +2369,8 @@ def check_fit_idempotent(name, estimator_orig):
-        if hasattr(estimator, method):
-            result[method] = getattr(estimator, method)(X_test)
+    result = {method: getattr(estimator, method)(X_test)
+              for method in check_methods if hasattr(estimator, method)}
```
I personally prefer the if on the next line

jnothman reviewed (commented):
As this is all about cosmetics, I've made some cosmetic comments :)

Have you checked docs/examples, where we really expect users to copy our coding style?

Do we have reason to believe that these changes are unlikely to make conflicts?

rth committed:
Minor fixes in examples

rth committed:
Joel's comments

rth committed:
Merge branch 'master' into use-list-comprehension

rth commented:
Thanks for the review. @jnothman I addressed your comments.

> Have you checked docs/examples, where we really expect users to copy our coding style?

Added some in https://github.com/scikit-learn/scikit-learn/pull/12668/commits/5a07c5cfc480a1013cb2f185435a603b991ef1cf, couldn't find that much more such cases.

> Do we have reason to believe that these changes are unlikely to make conflicts?

Well, this adds 48 lines and removes 85 lines -- that's nothing compared to hundreds / thousand lines changed in some recent PRs (e.g. dropping py2 or `assert_{true,false}`).

jnothman commented:
I'm not concerned by merge conflicts created through changes to tests.
Tests are mostly appended to...

rth commented:
I know, my point is that I don't feel we currently have an issue with merge conflicts, unlike with review bandwidth etc. And I would rather help resolve a few conflicts in PRs than not merge maintenance changes in general.

Also, this changes relatively few code lines and I don't think would create many conflicts.

jnothman commented:
tx @rth
